scripts/autolabel.lean

```diff
@@ -0,0 +1,286 @@
+  applicable.map (·.label) |>.qsort (· < ·)
+
+/-!
+Testing the functionality of the declarations defined in this script
+-/
+section Tests
+
+-- Test `FilePath.isPrefixOf`
+#guard ("Mathlib" / "Algebra" : FilePath).isPrefixOf ("Mathlib" / "Algebra" / "Basic.lean")
+
+-- Test `FilePath.isPrefixOf` does not trigger on partial prefixes
+#guard ! ("Mathlib" / "Algebra" : FilePath).isPrefixOf ("Mathlib" / "AlgebraicGeometry")
+
+#guard getMatchingLabels #[] == #[]
+-- Test default value for `label.dirs` works
+#guard getMatchingLabels #["Mathlib" / "SetTheory" / "ZFC"] == #["t-set-theory"]
+-- Test exclusion
+#guard getMatchingLabels #["Mathlib" / "Tactic"/ "Abel.lean"] == #["t-meta"]
+#guard getMatchingLabels #["Mathlib" / "Tactic"/ "Linter" / "Lint.lean"] == #["t-linter"]
+#guard getMatchingLabels #[
+  "Mathlib" / "Tactic"/ "Linter" / "Lint.lean",
+  "Mathlib" / "Tactic" / "Abel.lean" ] == #["t-linter", "t-meta"]
+
+/-- Testing function to ensure the labels defined in `mathlibLabels` cover all
+subfolders of `Mathlib/`. -/
+partial def findUncoveredPaths (path : FilePath) (exceptions : Array FilePath := #[]) :
+    IO <| Array FilePath := do
+  let mut notMatched : Array FilePath := #[]
+  -- all directories inside `path`
+  let subDirs ← (← path.readDir).map (·.path) |>.filterM (do FilePath.isDir ·)
+  for dir in subDirs do
+    -- if the sub directory is not matched by a label,
+    -- we go recursively into it
+    if (getMatchingLabels #[dir]).size == 0 then
+      notMatched := notMatched ++ (← findUncoveredPaths dir exceptions)
+  -- a directory should be flagged if none of its sub-directories is matched by a label
+  -- note: we assume here the base directory, i.e. "Mathlib" is never matched by a label,
+  -- therefore we skip this test.
+  if notMatched.size == subDirs.size then
+    if exceptions.contains path then
+      return #[]
+    else
+      return #[path]
+  else
+    return notMatched
+
+end Tests
+
+end AutoLabel
+
+open IO AutoLabel in
+
+/-- `args` is expected to have length 0 or 1, where the first argument is the PR number.
+
+If a PR number is provided, the script requires GitHub CLI `gh` to be installed in order
+to add the label to the PR.
+
+## Exit codes:
+
+- `0`: success
+- `1`: invalid arguments provided
+- `2`: invalid labels defined
+- `3`: ~labels do not cover all of `Mathlib/`~ (unused; only emitting warning)
+-/
+unsafe def main (args : List String): IO Unit := do
+  if args.length > 1 then
+    println s!"autolabel: invalid number of arguments ({args.length}). Please run without \
+    arguments or provide the target PR's number as single argument!"
+    IO.Process.exit 1
+  let prNumber? := args[0]?
+
+  -- test: validate that all paths in `mathlibLabels` actually exist
+  let mut valid := true
+  for label in mathlibLabels do
+    for dir in label.dirs do
+      unless ← FilePath.pathExists dir do
+        println s!"::error file=scripts/autolabel.lean,line=60::directory {dir} does not exist! \
```

Comment:
```suggestion
        println s!"::error file=scripts/autolabel.lean,line=60::directory {dir} does not exist! \
```
what does `line=60` refer to?

Reply:
`line=84` (updated) is supposed to be roughly the definition of `def mathlibLabels : Array Label`?

I don't know a way to find this automatically. Should I just do `line=1` instead?

Reply:
This may work (untested):
```lean
 line := ((declRangeExt.find? (← getEnv) `Autolabel.mathlibLabels).getD default).range.pos.line
```

Reply:
hmm, but I'm in `IO` and don't really have access to the `env`. Usually I would use `withImportModules` in this case, but that does not work since `script/autolabel.lean` is not importable.

Reply:
How about saying "directory `foo` in `mathlibLabels` does not exist"? This avoids line numbers at all, but is hopefully equally informative.

Reply:
done